distributed/active_memory_manager.py

```diff
@@ -0,0 +1,429 @@
+        replications and drops) will be returned.
+        """
+        if len(ts.who_has) < 2:
+            return None
+        if candidates is None:
+            candidates = ts.who_has
+        else:
+            candidates = ts.who_has & candidates
+        candidates = candidates - {waiter_ts.processing_on for waiter_ts in ts.waiters}
+        if not candidates:
+            return None
+        # id(ws) is there just to prevent WorkerState objects to be compared in the
+        # unlikely event that they have exactly the same amount of bytes allocated.
+        _, _, ws = max((self.workers_memory[ws], id(ws), ws) for ws in candidates)
+        return ws
+
+
+class ActiveMemoryManagerPlugin:
+    """Abstract class"""
+
+    manager: ActiveMemoryManager
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}()"
+
+    def run(
+        self,
+    ) -> """Generator[
+        tuple[str, TaskState, Optional[set[WorkerState]]],
+        Optional[WorkerState],
+        None
+    ]""":
+        """This method is invoked by the ActiveMemoryManager every few seconds.
+        It is an iterator that must emit one of the following:
+
+        - "replicate", <TaskState>, None
+        - "replicate", <TaskState>, {subset of potential workers to replicate to}
+        - "drop", <TaskState>, None
+        - "drop", <TaskState>, {subset of potential workers to drop from}
+
+        Each element yielded indicates the desire to create or destroy a single replica
+        of a key. Either the ActiveMemoryManager or the Worker may later decide to
+        disregard the request, e.g. because it would delete the last copy of a key or
+        because the key is currently needed on that worker.
+
+        You may optionally acquire the decision of which worker the key will be
+        replicated on or dropped from, as follows:
+
+        ```python
+        choice = (yield "replicate", ts, None)
+        ```
+        ``choice`` is either a WorkerState or None; the latter is if the
+        ActiveMemoryManager chose to disregard the request.
+
+        The current pending (accepted) commands can be inspected on
+        ``self.manager.pending``; this includes the commands previously yielded by this
+        same method.
+
+        The current memory usage on each worker, *downstream of all pending commands*,
+        can be inspected on ``self.manager.workers_memory``.
+        """
+        raise NotImplementedError("Virtual method")
+
+
+class ReplicateReduce(ActiveMemoryManagerPlugin):
+    """Make sure that we don't have more replicas than desired for every key; drop
+    excess replicas
+    """
+
+    def run(self):
+        # TODO this is O(n) to the total number of in-memory tasks on the cluster; it
+        #      could be made faster by automatically attaching it to a key when it goes
+        #      above one replica and detaching it when it goes below.
```

Comment:
putting a hook on additions of TaskManager.who_has would be a very simple design and it would also mean we can merge ReplicateReduce and ReplicateIncrease into a single plugin